### PR TITLE
Fix Deepspeed and lightning calling scheduler

### LINKED ISSUE #9771: DeepSpeed & Lightning are both calling the scheduler
## 🐛 Bug

A user in our slack channel noticed that the scheduler step is being called by DeepSpeed and Lightning. Lightning should not call this, and DeepSpeed should be in control.

### To Reproduce

This test fails!
```python
import os

import mock
import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 1, gamma=0.1)
        return {
            "optimizer": optimizer,
            "lr_scheduler": {
                "scheduler": scheduler,
            },
        }


@mock.patch('torch.optim.lr_scheduler.StepLR.step', autospec=True)
def test_deepspeed(mock_step):
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=0,
        num_sanity_val_steps=0,
        max_epochs=1,
        fast_dev_run=True,
        gpus=1,
        weights_summary=None,
        plugins='deepspeed'
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    # once before training, and once during training.
    assert mock_step.call_count == 2
```

### Expected behaviour

Above should succeed. I also see there may be an issue with `test` also calling the step function internally in DeepSpeed, but thats for a different issue.

cc @tchaton 

### Environment

* CUDA:
        - GPU:
                - NVIDIA GeForce RTX 3090
        - available:         True
        - version:           11.1
* Packages:
        - numpy:             1.21.2
        - pyTorch_debug:     False
        - pyTorch_version:   1.9.1
        - pytorch-lightning: 1.5.0dev
        - tqdm:              4.49.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.8

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Fixes #9771 

After speaking to @tchaton we decided to remove the scheduler responsibility from DeepSpeed, fully relying on Lightning to control when to call the lr scheduler step function. This also allows users to use epoch vs step calls which is not possible in DeepSpeed.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LGMT !

**Review by awaelchli — APPROVED** [review]

File: tests/plugins/test_deepspeed_plugin.py [hunk 1]
```diff
@@ -921,3 +921,47 @@ def test_dataloader(self):
...
+    set to step."""
+    _run_scheduler_test(mock_step, max_epoch=2, limit_train_batches=2, interval="step")
+
+
+@mock.patch("torch.optim.lr_scheduler.StepLR.step", autospec=True)
+@RunIf(min_gpus=1, deepspeed=True, special=True)
+def test_deepspeed_scheduler_step_count_epoch(mock_step):
+    """Test to ensure that the scheduler is called the correct amount of times during training when scheduler is
+    set to epoch."""
+    _run_scheduler_test(mock_step, max_epoch=2, limit_train_batches=2, interval="epoch")
```
Comment (rohitgr7) [inline tests/plugins/test_deepspeed_plugin.py:939]: maybe parametrize the interval here and put the below function logic inside the test?? else looks good!!

**Review by rohitgr7 — APPROVED** [review]

File: tests/plugins/test_deepspeed_plugin.py (hunk 1 above)
Comment (SeanNaren) [reply tests/plugins/test_deepspeed_plugin.py:939]: Oh sorry I should've mentioned, I couldn't make this work (probably because its a special test?) I had to split it up like this unfortunately :(

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 2]
```diff
@@ -441,7 +441,11 @@ def _initialize_deepspeed_train(self, model):
 
         # although we set these here, deepspeed manages the specific optimizer logic
         self.lightning_module.trainer.optimizers = [deepspeed_optimizer]
+
+        deepspeed_scheduler = model.lr_scheduler
         if deepspeed_scheduler is not None:
+            # disable deepspeed lr scheduling as lightning manages scheduling
+            model.lr_scheduler = None
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/deepspeed.py:448]: why this option instead of not passing it to `deepspeed.initialize` above?

Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:448]: The main reason is if the user has used the config file to determine what the optimizer/scheduler should be (this is the case if you're using the standard deepspeed recipe config and want to replicate something) then we still need to use deepspeed to select the appropriate lr scheduler if any
